kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java

```diff
@@ -109,9 +162,49 @@ private void addHandlers(SocketChannel ch, VirtualCluster virtualCluster) {
             pipeline.addLast(new KafkaAuthnHandler(ch, authnHandlers));
         }
 
-        var netFilter = new UpstreamBrokerAddressCachingNetFilter(config, virtualCluster);
+        var frontendHandler = new KafkaProxyFrontendHandler(context -> {
+            var filterChainFactory = new FilterChainFactory(config, virtualCluster.getClusterEndpointProvider());
+
+            var filters = new ArrayList<>(Arrays.stream(filterChainFactory.createFilters()).toList());
+
+            // Add a filter to the *end of the chain* that gathers the true nodeId/upstream broker mapping.
+            filters.add((MetadataResponseFilter) (header, response, filterContext) -> {
+                response.brokers().forEach(b -> {
+                    var replacement = new HostPort(b.host(), b.port());
+                    var existing = virtualCluster.updateUpstreamClusterAddressForNode(b.nodeId(), replacement);
+                    if (!replacement.equals(existing)) {
+                        LOGGER.info("Got upstream for broker {} : {}", b.nodeId(), replacement);
+                    }
+                });
+                filterContext.forwardResponse(response);
+            });
+
+            var targetBootstrapServers = virtualCluster.targetCluster().bootstrapServers();
+            var targetBootstrapServersParts = targetBootstrapServers.split(",");
+            var targetClusterBootstrap = HostPort.parse(targetBootstrapServersParts[0]);
+
+            HostPort target;
+            if (binding instanceof VirtualClusterBrokerBinding brokerBinding) {
```

Comment:
could we refactor this to instead ask something like `brokerBinding.getTargetHostPost()`, I think the bindings have everything they need to do this work without an instanceof.